pandas/core/algorithms.py

```diff
@@ -1891,3 +1893,138 @@ def union_with_duplicates(lvals: ArrayLike, rvals: ArrayLike) -> ArrayLike:
     for i, value in enumerate(unique_array):
         indexer += [i] * int(max(l_count[value], r_count[value]))
     return unique_array.take(indexer)
+
+
+# ------ #
+# sample #
+# ------ #
+
+
+def preprocess_weights(obj: FrameOrSeries, weights, axis: int) -> np.ndarray:
```

Comment:
This was pretty much moved as is, with the only change being to convert to an `ndarray` earlier for better performance on later validation steps

Reply:
im really not a fan of Series/DataFrame methods living in this file.  are there any other natural homes for this?

Reply:
`util/_validators` is almost a good fit, but while this validates, it also returns a modified input. Could go in `core/common`? But still learning where everything lives, happy to move if anyone has a better location

Reply:
looks like this function is the only one that really depends on Series/DataFrame; could it just stay inside the NDFrame method?  the others could go in e.g. core.array_algos.sample

Reply:
That would be nicer, but the issue is that the weights processing also needs to be called from groupby sample as well.

Reply:
Another option would be to implement something like a core.methods directory for Series/DataFrame methods that have been refactored to their own files (e.g. describe).  I think algos.SelectN might make sense in something like that

Reply:
I like that idea - have moved to core/sample.py (on same level as describe) for now. If others like this organization, I can follow up moving describe and sample (and maybe others like you mention) to core/methods.